R/unit.R

```diff
@@ -407,119 +352,31 @@ print.unit <- function(x, ...) {
-  }
-  repSummaryUnit <- function(x, n) {
-      val <- get(x$fname)(x$arg1)
-      newUnits <- lapply(integer(n), function(z) val)
-      class(newUnits) <- c("unit.list", "unit")
-      newUnits
-  }
-
-  switch(x$fname,
-         "+"=`[`(x$arg1, (index - 1L) %% this.length + 1L, top=FALSE) +
-             `[`(x$arg2, (index - 1L) %% this.length + 1L, top=FALSE),
-         "-"=`[`(x$arg1, (index - 1L) %% this.length + 1L, top=FALSE) -
-             `[`(x$arg2, (index - 1L) %% this.length + 1L, top=FALSE),
-         # Recycle multiplier if necessary
-         "*"=`[`(x$arg1, (index - 1L) %% length(x$arg1) + 1L) *
-             `[`(x$arg2, (index - 1L) %% this.length + 1L, top=FALSE),
-         "min"=repSummaryUnit(x, length(index)),
-         "max"=repSummaryUnit(x, length(index)),
-         "sum"=repSummaryUnit(x, length(index)))
-}
-
-`[.unit.list` <- function(x, index, top=TRUE, ...) {
-  this.length <- length(x)
-  if (is.logical(index))
-    index <- which(index)
-  else { # Allow for negative integer index
-    if (any(index < 0)) {
-      if (any(index > 0))
-        stop("cannot mix signs of indices")
-      else
-        index <- (1L:this.length)[index]
-    }
-    if (top && any(index > this.length))
-      stop("index out of bounds (unit list subsetting)")
+`[.unit` <- function(x, index, top = TRUE) {
+  attr <- attributes(x)
+  x <- unclass(x)
+  n <- length(x)
+  if (is.numeric(index) && any(index > n)) {
+    if (top) stop('index out of bounds ("unit" subsetting)', call. = FALSE)
+    index <- (seq_len(n)[index] - 1L) %% n + 1L
   }
-  structure(class = class(x),
-            unclass(x)[(index - 1L) %% this.length + 1L])
+  x <- x[index]
+  `attributes<-`(x, attr)
 }
-
-# `[<-.unit` methods
-#
-# The basic approach is to convert everything to a unit.list,
-# unclass (so everything is list), rely on list subassignment, reclass
-
 `[<-.unit` <- function(x, i, value) {
-    if (!is.unit(value))
-        stop("Value being assigned must be a unit")
-    valueList <- unclass(unit.list(value))
-    xList <- unclass(unit.list(x))
-    xList[i] <- valueList
-    class(xList) <- c("unit.list", "unit")
-    xList
-}
-
-#########################
-# str() method
-#########################
-
-# Should work fine on atomic units and on unit.list
-# The problem arises with unit.arithmetic, which are stored as lists
-# but act like vectors
-# (e.g., report length greater than number of list components)
-str.unit.arithmetic <- function(object, ...) {
-    cat("Class 'unit.arithmetic' [1:", length(object), "] ", sep="")
-    str(unclass(object), ...)
+  if (!is.unit(value)) stop('value must be a unit object')
+  attr <- attributes(x)
+  if (is.simpleUnit(x)) {
+  	if (!(is.simpleUnit(value) && attr(x, 'unit') == attr(value, 'unit'))) {
+  	  x <- as.unit(x)
+  	  value <- as.unit(value)
+  	}
+  } else {
+  	value <- as.unit(value)
+  }
+  x <- unclass(x)
+  x[i] <- value
+  `attributes<-`(x, attr)
```

Comment:
This is wrong because it resets the class attribute - so a "simpleUnit" coming in gets its class set BACK to "simpleUnit" even though it has been changed to "unit" by as.unit().  Simple demo (from example(unit)) ...
```
x <- unit(1:5, "npc")
x[2:4]
x[2:4] <- unit(1, "mm")
```
Fix is this ... ?
```
  attr <- attributes(x)
  simpleResult <- FALSE
  if (is.simpleUnit(x)) {
  	if (!(is.simpleUnit(value) && attr(x, 'unit') == attr(value, 'unit'))) {
  	  x <- as.unit(x)
  	  value <- as.unit(value)
  	} else {
          simpleResult <- TRUE
        }
  } else {
  	value <- as.unit(value)
  }
  x <- unclass(x)
  x[i] <- value
  if (simpleResult) {
    `attributes<-`(x, attr)```
  } else {
    class(x) <- "unit"
  }
  x
```

Reply:
Thanks for catching that!